### PR TITLE
findContours: integrate workaround to allow contours detection on image border

### LINKED ISSUE #4374: Strange behavior of cv::findContours at image boundary
Transferred from http://code.opencv.org/issues/2045

```
|| Changlin Hsieh on 2012-06-13 10:24
|| Priority: Low
|| Affected: None
|| Category: imgproc, video
|| Tracker: Feature
|| Difficulty: None
|| PR: None
|| Platform: None / None
```
## Strange behavior of cv::findContours at image boundary

```
The contours return by findContours on the boundary will be 1 or width-2 or height-2, I think this is really strange.
Is there any reason to do this or I misunderstand the using of findContour?

This method will modify input segment is also a strange behavior.
Although this can be solve by clone an input before use this method, but I think this should be done in this method.
A user will not expect a broken segment after find contour and this broken segment just useless later.
```
##### Kirill Kornyakov on 2012-06-15 07:21

```
Sorry, but it is hard to understand you... Let me rephrase, and check if I get you right.
# The first problem is that you think that something wrong with contours. It seems that their width and height is 1 or 2, and you don't like it... Is it so? *Could you please elaborate on what exactly is wrong with contours, and how you see the correct behavior?*
# You don't like that input segments are modified within function.
-   Assignee changed from Vadim Pisarevsky to Changlin Hsieh
```
##### Changlin Hsieh on 2012-06-15 07:47

```
What I think of question 1 is the contour found by this function at image boundary will like (1,1), (1,2), (1,3), ..., (1,n), but not (0,1), (0,2), (0,3), ..., (0,n).
Because of this, if I draw the contour on the image it will remain something outer the boundary like the "contour.png".

For question 2, yes I don't like this method modify the segments
```
##### Andrey Kamaev on 2012-06-18 08:55

```
-   Assignee deleted (Changlin Hsieh)
```

### LINKED ISSUE #4988: Off-by-one in findContours
Transferred from http://code.opencv.org/issues/4329

```
|| Michele Santullo on 2015-05-11 14:49
|| Priority: Normal
|| Affected: 2.4.9 (latest release)
|| Category: imgproc, video
|| Tracker: Bug
|| Difficulty: 
|| PR: 
|| Platform: x86 / Windows
```
## Off-by-one in findContours

```
I have the following mask: https://imgur.com/oy4xuLX (in case the link becomes invalid, it's a black 512x424 png with a white rectangle from x-y 128-0 to 383-365 inclusive).
I want to write a function to retrieve the total bounding rect around all of the white areas (in this case just one). I have the following c++ code:

  cv::Rect get_useful_area (const cv::Mat& parMask) {
    std::vector<std::vector<cv::Point>> contours;
    std::vector<cv::Vec4i> hierarchy;
    cv::findContours(parMask, contours, hierarchy, CV_RETR_CCOMP, CV_CHAIN_APPROX_SIMPLE);
    cv::Rect bounds;
    if (not contours.empty())
      bounds = cv::boundingRect(contours[0]);
      for (std::size_t z = 1; z < contours.size(); ++z) {
      bounds |= cv::boundingRect(contours[z]);
    }
    return bounds;
  }
  void Mask::load (const std::string& parPath) {
    m_mask = cv::imread(parPath, CV_LOAD_IMAGE_GRAYSCALE);
    m_right = m_bottom = 0;
    m_left = m_mask.cols;
    m_top = m_mask.rows;
    for (int y = 0; y < m_mask.rows - 1; ++y){
      for (int x = 0; x < m_mask.cols - 1; ++x){
        const unsigned char val = *m_mask.ptr<unsigned char>(y, x);
        if (val) {
          m_left = std::min(m_left, x);
          m_top = std::min(m_top, y);
          m_right = std::max(m_right, x + 1);
          m_bottom = std::max(m_bottom, y + 1);
        }
      }
    }
    {
      const auto bounds = get_useful_area(m_mask);
      vaeAssert(bounds.x == m_left);
      vaeAssert(bounds.y == m_top);
      vaeAssert(bounds.x + bounds.width == m_right);
      vaeAssert(bounds.y + bounds.height == m_bottom);
    }
  }

With the above image as the input the second assertion fails. In the debugger I get:
bounds.x = 128
bounds.y = 1
bounds.width = 256
bounds.height = 365
m_left = 128
m_top = 0
m_right = 384
m_bottom = 366

The sum bounds.y+bounds.height happens to be correct (vertically, the white rectangle goes from row 0 to row 365 inclusive, so it's 366 pixels high), however both y and height are wrong.
```
##### Vadim Pisarevsky on 2015-05-21 19:31

```
-   Category set to imgproc, video
```
##### enpe enpe on 2015-07-12 19:51

```
Added linked image as an attachment (in case the links dies, as the OP already noted).
-   File oy4xuLX_-_Imgur.png added
```

### PR DESCRIPTION
**Pull request by sovrasov**
Resolves #4374
Resolves #4988

### CODE REVIEW

File: modules/imgproc/src/contours.cpp [hunk 1]
```diff
@@ -1878,13 +1878,14 @@ void cv::findContours( InputOutputArray _image, OutputArrayOfArrays _contours,
 
     CV_Assert(_contours.empty() || (_contours.channels() == 2 && _contours.depth() == CV_32S));
 
-    Mat image = _image.getMat();
+    Mat image;
+    cv::copyMakeBorder(_image.getMat(), image, 1, 1, 1, 1, cv::BORDER_CONSTANT, cv::Scalar(0));
```
Comment (alalek) [inline modules/imgproc/src/contours.cpp]: `_image.getMat()` -> `_image`

**Comment by mschoeneck** [comment]
There should be already exist a routine that copies the image with a zero border starting at line 313 in the code. Please check this for the problem.

    /* make zero borders */
    int esz = CV_ELEM_SIZE(mat->type);
    memset( img, 0, size.width*esz );
    memset( img + static_cast<size_t>(step) * (size.height - 1), 0, size.width*esz );

    img += step;
    for( int y = 1; y < size.height - 1; y++, img += step )
    {
        for( int k = 0; k < esz; k++ )
            img[k] = img[(size.width - 1)*esz + k] = (schar)0;
    }

**Comment by catree** [comment]
@sovrasov Looks like the `/* make zero borders */` code is useless now as we already created a bigger image (width+2, height+2) with a 1-pixel border filled with 0's.

The [documentation](http://docs.opencv.org/3.1.0/d3/dc0/group__imgproc__shape.html#ga17ed9f5d79ae97bd4c7cf18403e1689a) should also be updated to remove the comment about the 1-pixel border.

**Comment by alalek** [comment]
I believe regression test should help here.

**Comment by sovrasov** [comment]
@catree unfortunately this will break C variant of the function, which is used in chessboard detector.

**Comment by mschoeneck** [comment]
I would also like to point out, that there is another algorithm besides the Suzuki which does not need an image with a zero border (CV_LINKRUN). For this one copyMakeBorder is just unnecessary overhead.

In this context it would be possible to remove the threshold function from cvStartFindContours and put it into the findContours function. Also it would be possible to make use of OpenCL for them.

A nice to have would be to pass a flag for not returning a chain code image with an extra function header. This avoids the need of clone the input image in many cases.

File: modules/imgproc/include/opencv2/imgproc.hpp [hunk 2]
```diff
@@ -3732,21 +3732,16 @@ CV_EXPORTS_W int connectedComponentsWithStats(InputArray image, OutputArray labe
...
 
-@note Source image is modified by this function. Also, the function does not take into account
-1-pixel border of the image (it's filled with 0's and used for neighbor analysis in the algorithm),
-therefore the contours touching the image border will be clipped.
-
 @param image Source, an 8-bit single-channel image. Non-zero pixels are treated as 1's. Zero
 pixels remain 0's, so the image is treated as binary . You can use compare , inRange , threshold ,
 adaptiveThreshold , Canny , and others to create a binary image out of a grayscale or color one.
-The function modifies the image while extracting the contours. If mode equals to RETR_CCOMP
-or RETR_FLOODFILL, the input can also be a 32-bit integer image of labels (CV_32SC1).
```
Comment (catree) [inline modules/imgproc/include/opencv2/imgproc.hpp]: Is there a reason why the comment:

> If mode equals to RETR_CCOMP or RETR_FLOODFILL, the input can also be a 32-bit integer image of labels (CV_32SC1).

has been removed?

File: modules/imgproc/include/opencv2/imgproc/imgproc_c.h [hunk 3]
```diff
@@ -435,12 +441,13 @@ CVAPI(int)  cvFindContours( CvArr* image, CvMemStorage* storage, CvSeq** first_c
                             int method CV_DEFAULT(CV_CHAIN_APPROX_SIMPLE),
                             CvPoint offset CV_DEFAULT(cvPoint(0,0)));
 
-/** @brief Initializes contour retrieving process.
+CVAPI(int)  cvFindContours2( CvArr* image, CvMemStorage* storage, CvSeq** first_contour,
+                            int header_size CV_DEFAULT(sizeof(CvContour)),
+                            int mode CV_DEFAULT(CV_RETR_LIST),
+                            int method CV_DEFAULT(CV_CHAIN_APPROX_SIMPLE),
+                            CvPoint offset CV_DEFAULT(cvPoint(0,0)), int needFillBorder CV_DEFAULT(1));
```
Comment (alalek) [inline modules/imgproc/include/opencv2/imgproc/imgproc_c.h]: My suggestion is to make these functions internal only (use 'static' in .cpp file).
BTW, You may use "Impl" suffix instead of "2".

**Comment by alalek** [comment]
:+1:
